### PR TITLE
LibGfx/JPEG: Reduce memory usage by almost 50% :^)

### PR DESCRIPTION
**Pull request by LucasChollet**
~Based on #17907 and #17786, only the 5 lasts commits are original.~

This is ready to merge!

I simplified this PR as the second part need more thought/work.

Master:
157,0 Mo
Completed benchmark 'small_image' on average in 0.0±0.0ms (min=0ms, max=0ms, total=0ms)
Completed benchmark 'big_image' on average in 614.2±7.5ms (min=597ms, max=645ms, total=61424ms)
Completed benchmark 'rgb_image' on average in 6.0±0.3ms (min=6ms, max=9ms, total=610ms)
Completed benchmark 'several_scans' on average in 5.2±0.5ms (min=5ms, max=7ms, total=528ms)

[LibGfx/JPEG: Use a smaller type to store coefficients](https://github.com/SerenityOS/serenity/pull/18077/commits/a28821b6307b0e8c8b841d278f6fe55d856f62cb):
84,8 Mo
Completed benchmark 'small_image' on average in 0.0±0.0ms (min=0ms, max=0ms, total=0ms)
Completed benchmark 'big_image' on average in 538.3±19.5ms (min=517ms, max=628ms, total=53835ms)
Completed benchmark 'rgb_image' on average in 6.0±0.1ms (min=6ms, max=7ms, total=604ms)
Completed benchmark 'several_scans' on average in 5.0±0.1ms (min=5ms, max=6ms, total=502ms)

### CODE REVIEW

**Comment by nico** [comment]
The actual change lgtm.

Notes on the benchmark:

1. Adding a 5.8 MiB binary file to the repo is maybe not ideal (?)
2. It's weird to me that the images are loaded at static initialization time instead of in the test itself. It doesn't even separate file system access from the test – all disk i/o will happen when the data is paged in on first access.

Neither of these are necessarily blocking (up to maintainers), but it seemed a bit surprising to me.

**Comment by LucasChollet** [comment]
> Adding a 5.8 MiB binary file to the repo is maybe not ideal (?)

I'm not very happy with that either but I decided to let a maintainer take the decision. However, I think that it is important to benchmark the decoder on bigger images.

> It's weird to me that the images are loaded at static initialization time instead of in the test itself. It doesn't even separate file system access from the test – all disk i/o will happen when the data is paged in on first access.

Oops, my bad mapped files were an error. Loading at static time was by design tho. I want to test the decoder regardless of I/O speed.

**Comment by nico** [comment]
Maybe consider pulling out the actual memory savings into their own PR – I'd imagine that part will be merged immediately :)

File: Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp [hunk 1]
```diff
@@ -102,18 +102,18 @@ using Marker = u16;
  */
 struct Macroblock {
     union {
-        i32 y[64] = { 0 };
-        i32 r[64];
+        i16 y[64] = { 0 };
```
Comment (nico) [inline Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp:105]: do you happen to know where the ITU doc says that these are signed?

**Review by nico — APPROVED** [review]
lg!

**Comment by nico** [comment]
(note to self: tests now in #18162)

File: Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp (hunk 1 above)
Comment (LucasChollet) [reply Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp:105]: > A.1.2 Sample precision
A sample is an integer with precision P bits, with any value in the range 0 through 2^P – 1. All samples of all components within an image shall have the same precision P. Restrictions on the value of P depend on the mode of operation, as specified in B.2 to B.7.

Which means that the input data is unsigned.

> A.3.1 Level shift
Before a non-differential frame encoding process computes the FDCT for a block of source image samples, the samples shall be level shifted to a signed representation by subtracting 2(P – 1), where P is the precision parameter specified in B.2.2. Thus, when P = 8, the level shift is by 128; when P = 12, the level shift is by 2048.

This makes values be in $\mathbb {Z}$.

> B.2.2 Frame header syntax
P: Sample precision – Specifies the precision in bits for the samples of the components in the frame.

**Comment by LucasChollet** [comment]
Hold on! The recently added 4th component isn't an i16.

**Pull request merged**